core/include/scipp/core/multi_index.h

```diff
@@ -30,276 +47,332 @@ auto get_slice_dim(const T &param, const Ts &... params) {
+                                                 params.strides()...);
+
+    m_buffer =
+        std::make_unique<scipp::index[]>(detail::get_buffer_size<N>(m_ndim));
+    // using manual size here because the max in get_* would overflow the buffer
+    std::copy(inner_strides_buffer, inner_strides_buffer + N * m_inner_ndim,
+              &stride(0, 0));
+    std::copy(bin_strides_buffer,
+              bin_strides_buffer + N * (m_ndim - m_inner_ndim),
+              &stride(m_inner_ndim, 0));
+    std::copy(inner_shape_buffer, inner_shape_buffer + m_inner_ndim,
+              shape_it());
+    std::copy(bin_shape_buffer, bin_shape_buffer + (m_ndim - m_inner_ndim),
+              shape_it(m_inner_ndim));
+
+    // NOLINTNEXTLINE(cppcoreguidelines-prefer-member-initializer)
+    m_bin_stride = inner_dims.offset(slice_dim);
+    // NOLINTNEXTLINE(cppcoreguidelines-prefer-member-initializer)
+    m_nested_dim_index = m_inner_ndim - inner_dims.index(slice_dim) - 1;
+
+    if (bin_dims.volume() == 0) {
+      return; // Operands are empty, there are no bins to load.
     }
-    copy_strides<N>(m_stride, nestedDims.ndim(),
-                    std::index_sequence_for<Params...>(),
-                    params.bucketParams() ? Strides{nestedDims} : Strides{}...);
-    std::array<std::array<scipp::index, NDIM_MAX>, N> binStrides;
-    copy_strides<N>(binStrides, m_ndim - m_ndim_nested,
-                    std::index_sequence_for<Params...>(), params.strides()...);
     for (scipp::index data = 0; data < N; ++data) {
-      for (scipp::index d = 0; d < NDIM_MAX - m_ndim_nested; ++d)
-        m_stride[data][m_ndim_nested + d] = binStrides[data][d];
       load_bin_params(data);
     }
-    if (m_shape[m_nested_dim_index] == 0)
+    if (shape(m_nested_dim_index) == 0)
       seek_bin();
   }
 
-  constexpr void load_bin_params(const scipp::index i) noexcept {
-    if (m_coord[m_ndim] != 0 || !m_bin[i].is_binned())
-      return; // at end or dense
-    // All bins are guaranteed to have the same size.
-    // Use common m_shape and m_nested_stride for all.
-    const auto [begin, end] = m_bin[i].m_indices[m_bin[i].m_bin_index];
-    m_shape[m_nested_dim_index] = end - begin;
-    m_data_index[i] = m_nested_stride * begin;
+public:
+  MultiIndex(const MultiIndex &other)
+      : m_buffer{other.copy_buffer()}, m_data_index{other.m_data_index},
+        m_inner_ndim{other.m_inner_ndim}, m_ndim{other.m_ndim},
+        m_bin_stride{other.m_bin_stride},
+        m_nested_dim_index{other.m_nested_dim_index}, m_bin{other.m_bin} {}
+
+  MultiIndex(MultiIndex &&) noexcept = default;
+
+  MultiIndex &operator=(const MultiIndex &other) {
+    m_buffer = other.copy_buffer();
+    m_data_index = other.m_data_index;
+    m_inner_ndim = other.m_inner_ndim;
+    m_ndim = other.m_ndim;
+    m_bin_stride = other.m_bin_stride;
+    m_nested_dim_index = other.m_nested_dim_index;
+    m_bin = other.m_bin;
+    return *this;
```

Comment:
More reasons against the heap alloc'ed buffer.... too much code that can break or have bugs.

Reply:
Or just use a copyable array type instead of `unique_ptr`. Then the defaulted functions would work.